R/CompDbSource.R

```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
+        if (debug) handle <- verbose()
+        else handle <- NULL
+        out <- vector("list", length(query))
+		names(out) <- query$peak_id
+        restQuery <- .param_to_list(param)
 		for (i in seq_along(query)) {
-			## get peaks data from query[i] as list/character string for MetFrag (assuming single peaks table per spectra - is it always valid?!)	
-			peaks <- peaksData(query[i])[[1]]
-			## We allow either 'minFragCount' number of fragment peaks in spectra (under important uncontrolled auumption of proper data pretreatment), OR 
-			##	a 
-			if (nrow(peaks) < mfParam@minFragCount & nrow(peaks) > 0) {
-				## In cases of very few fragment peaks, the MetFrag server is 'timing-out' due (probably) to the large search space -
-				##	thus, I think the it's better to send the query with an empty fragment spectra and at least get the basic initial search matrix.
-				peaks <- peaks[-c(1:nrow(peaks)),]
-				## ...the other option is to ignor these calls - leaving the input as-is and returning empty results:
+            qi <- query[i]
+			peaks <- peaksData(qi)[[1L]]
+			if (nrow(peaks) < param@minFragCount) {
+				## In cases of very few fragment peaks, the MetFrag server is
+                ## 'timing-out' due (probably) to the large search space -
+				##	thus, I think the it's better to send the query with an
+                ## empty fragment spectra and at least get the basic initial
+                ## search matrix.
+				## ...the other option is to ignor these calls - leaving the
+                ## input as-is and returning empty results:
 				# out[[i]] = list(
 					# mfResults = NULL,
 					# mfServerStatus = "NA"
 				# )
-				# next()				
-			}
-			
-			peakListString <- paste(apply(peaks, 1, function (x){ paste(x[1],x[2],sep = '_') }),collapse = ';')
-
-			mz <- spectraData(query[i])$precursorMz
-
-			chkAdducts <- validAdducts(query,param)	
-			
-			if (!all(chkAdducts)) { 
-				stop("Some spectra precursor adducts are not legal MetFrag adduct types: Use function 'validAdducts(MetFragParam())' to print a list of valid adduct names. 'matchSpectra' stopping.") 
-			}
-							
-			adduct <- spectraData(query)$precursorAdduct[i]
-			## All REST calls will be done inside this loop here:
-			out[[i]] <- lapply(1:length(adduct),function (i) {
-				neutralPrecursorMass <- MetaboCoreUtils::mz2mass(mz, adduct[i])
-				restQuery <- sapply(slotNames(param)[-which(slotNames(param) %in% c("validAdducts","THRESHFUN"))], function(x) { 
-					slot(param,x)
-				})
-				restQuery <- c(
-					restQuery,	
-					"neutralprecursormass" = as.character(neutralPrecursorMass),
-					"precursoriontype" = as.character(adduct),
-					"peakliststring" = peakListString
+				# next()
+                peakListString <- ""
+			} else
+                peakListString <- .peaks_to_metfrag_string(peaks)
+			pmz <- qi$precursorMz
+			adduct <- unlist(qi$precursorAdduct)
+			out[[i]] <- do.call(rbind, lapply(adduct, function(a) {
```

Comment:
in contrast to the initial version I suggest to return a `data.frame` as a result and not a nested list. We add the `query_idx` and `adduct` to the `data.frame`, thus it's easier to identify which matches belong to each query. And we don't have to iterate/loop through the nested list later to create the result object (which will be anyway a `data.frame`).